cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/SourceControlManager.java

```diff
@@ -0,0 +1,253 @@
+  private final SourceControlContext context;
+  private final AuthStrategy authStrategy;
+  private final File localRepoDirectory;
+  private final File validationRepoDirectory;
+  private final int gitCommandTimeoutSeconds;
+  private Git git;
+
+  public SourceControlManager(SourceControlContext context) throws AuthStrategyNotFoundException {
+    this.context = context;
+    this.authStrategy = getAuthStrategy();
+    this.localRepoDirectory = context.getRepoBasePath().resolve(localRepositoryDirectoryName).toFile();
+    this.validationRepoDirectory = context.getRepoBasePath().resolve(validationRepositoryDirectoryName).toFile();
+    this.gitCommandTimeoutSeconds = context.getGitCommandTimeoutSeconds();
+  }
+
+  private <C extends TransportCommand> C getAuthenticatedCommandWithTimeout(C command) throws AuthenticationException {
+    command.setCredentialsProvider(authStrategy.getCredentialProvider(context));
+    Optional<TransportConfigCallback> transport = authStrategy.getTransportConfigCallback(context);
+    transport.ifPresent(command::setTransportConfigCallback);
+    command.setTimeout(gitCommandTimeoutSeconds);
+    return command;
+  }
+
+  /**
+   * Initialises the Git repository by cloning remote.
+   *
+   * @throws GitAPIException         when a Git operation fails.
+   * @throws AuthenticationException when correct authentication credentials are not available.
+   * @throws IOException             when file or network I/O fails.
+   */
+  public void initialise() throws GitAPIException, AuthenticationException, IOException {
+    if (git != null) {
+      return;
+    }
+    // Clean up the directory if it already exists.
+    deleteIfRequired(localRepoDirectory);
+    CloneCommand command =
+      Git.cloneRepository().setURI(context.getRepositoryConfig().getDefaultBranch()).setDirectory(localRepoDirectory);
+
+    String branch = getBranchRefName(context.getRepositoryConfig().getDefaultBranch());
+    if (branch != null) {
+      command.setBranchesToClone(Collections.singleton((branch))).setBranch(branch);
+    }
+    command = getAuthenticatedCommandWithTimeout(command);
+    command.call();
+  }
+
+  // Returns the ref name for the branch or Null if the branch name is null.
+  @Nullable
+  private String getBranchRefName(@Nullable String branch) {
+    if (Strings.isNullOrEmpty(branch)) {
+      return null;
+    }
+    return "refs/heads/" + branch;
+  }
+
+  /**
+   * Returns an {@link AuthStrategy} for the given Git repository provider and auth type.
+   *
+   * @return an instance of {@link  AuthStrategy}.
+   * @throws AuthStrategyNotFoundException when the corresponding {@link AuthStrategy} is not found.
+   */
+  private AuthStrategy getAuthStrategy() throws AuthStrategyNotFoundException {
+    RepositoryConfig config = this.context.getRepositoryConfig();
+    if (config.getProvider() == Provider.GITHUB) {
+      if (config.getAuth().getType() == AuthType.PAT) {
```

Comment:
Yes, we plan to have username password post MVP. Since Git doesn't have the concept of token based auth, all providers accept the token in the username or password field. CDAP will not know the format of all providers, but users will still use username/password auth and lookup the docs for their git hosting service to authenticate with their preferred repo hosting service.